### PR TITLE
TF: XLA stable softmax

### LINKED ISSUE #16838: TF: XLA model output differs when certain outputs are passed
Depending on the passed inputs, the output of an XLA-compiled model may significantly differ from its non-XLA counterpart. This suggests we should add tests for XLA-output equivalence, just like we do with e.g. PT-TF, as it is not guaranteed.

At the moment, this blocks further developments in `generate()` (can't reliably reproduce non-XLA results with XLA). I will assess this problem for T5 (first model where I've noticed this), then check whether it is present for other key models, and finally add equivalence tests.

cc @patrickvonplaten @Rocketknight1 (feel free to pitch in with ideas and suggestions)

____________________________________________________
Example for reproducibility (updated: assert diff < x-> print diff):
```python
import tensorflow as tf
from transformers import TFT5ForConditionalGeneration, T5Tokenizer


tokenizer = T5Tokenizer.from_pretrained("t5-base")
model = TFT5ForConditionalGeneration.from_pretrained("t5-base")
model_xla = tf.function(model, jit_compile=True)
pad_token_id = model.config.pad_token_id

sentence_1 = "Translate English to German: I have a cat, two dogs, three horses, and four birds."
sentence_2 = "Translate English to German: I have a cat, two dogs, and three horses."

ids_single = tokenizer([sentence_1], return_tensors="tf", padding=True).input_ids
decoder_ids_single = tf.zeros((1, 1), dtype=tf.int32)
attention_single = tf.cast(tf.math.not_equal(ids_single, pad_token_id), dtype=tf.int32)  # as computed in generate

ids_pair = tokenizer([sentence_1, sentence_2], return_tensors="tf", padding=True).input_ids
decoder_ids_pair = tf.zeros((2, 1), dtype=tf.int32)
attention_pair = tf.cast(tf.math.not_equal(ids_pair, pad_token_id), dtype=tf.int32)  # as computed in generate

# case 1: with batch size = 1 and NO attention mask, XLA and non-XLA match
outputs = model(input_ids=ids_single, decoder_input_ids=decoder_ids_single)
outputs_xla = model_xla(input_ids=ids_single, decoder_input_ids=decoder_ids_single)
print(tf.math.reduce_max(tf.math.abs(outputs.logits - outputs_xla.logits)).numpy())

# case 2: with batch size > 1 and NO attention mask, XLA and non-XLA match
outputs = model(input_ids=ids_pair, decoder_input_ids=decoder_ids_pair)
outputs_xla = model_xla(input_ids=ids_pair, decoder_input_ids=decoder_ids_pair)
print(tf.math.reduce_max(tf.math.abs(outputs.logits - outputs_xla.logits)).numpy())

# case 3 FAILING: with batch size = 1 and attention mask, XLA and non-XLA match
outputs = model(input_ids=ids_single, decoder_input_ids=decoder_ids_single, attention_mask=attention_single)
outputs_xla = model_xla(input_ids=ids_single, decoder_input_ids=decoder_ids_single, attention_mask=attention_single)
print(tf.math.reduce_max(tf.math.abs(outputs.logits - outputs_xla.logits)).numpy())

# case 4 FAILING: with batch size < 1 and attention mask, XLA and non-XLA match
outputs = model(input_ids=ids_pair, decoder_input_ids=decoder_ids_pair, attention_mask=attention_pair)
outputs_xla = model_xla(input_ids=ids_pair, decoder_input_ids=decoder_ids_pair, attention_mask=attention_pair)
print(tf.math.reduce_max(tf.math.abs(outputs.logits - outputs_xla.logits)).numpy())
```

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

As discussed in the thread about XLA problems (https://github.com/huggingface/transformers/issues/16838), this PR adds a stable wrapper for the softmax operation, and replaces `tf.nn.softmax` by the wrapped function.

This PR:
- Adds the wrapped softmax, named `stable_softmax`, in `tf_utils.py`. Its docstring includes why it is needed and why the new operation is valid;
- Adds tests to the wrapped softmax, including XLA tests;
- Replaces `tf.nn.softmax` by `stable_softmax` everywhere except in the doctests (I think it overcomplicates the examples, and no XLA should be needed there);
- Removes the `skipIf` for XLA tests, as they can now be successfully executed in a CPU.

Closes #16838

### CODE REVIEW

File: src/transformers/tf_utils.py [hunk 1]
```diff
@@ -44,3 +44,22 @@ def shape_list(tensor: Union[tf.Tensor, np.ndarray]) -> List[int]:
     static = tensor.shape.as_list()
 
     return [dynamic[i] if s is None else s for i, s in enumerate(static)]
+
+
+def stable_softmax(logits: tf.Tensor, axis: Optional[int] = None, name: Optional[str] = None) -> tf.Tensor:
+    """
+    Stable wrapper that returns the same output as `tf.nn.softmax`, but that works reliably with XLA on CPU. It is
+    meant as a workaround for the following issue (https://github.com/tensorflow/tensorflow/issues/55682), and will be
```
Comment (sgugger) [inline src/transformers/tf_utils.py:52]:
```suggestion
    meant as a workaround for the [following issue](https://github.com/tensorflow/tensorflow/issues/55682), and will be
```

File: src/transformers/tf_utils.py [hunk 2]
```diff
@@ -44,3 +44,22 @@ def shape_list(tensor: Union[tf.Tensor, np.ndarray]) -> List[int]:
...
+    removed after it gets fixed. The arguments and outputs are the same as `tf.nn.softmax`, and relies on the fact that
+    softmax(x) = softmax(x + c) (see https://ogunlao.github.io/2020/04/26/you_dont_really_know_softmax.html).
```
Comment (sgugger) [inline src/transformers/tf_utils.py:54]:
```suggestion
    `softmax(x) = softmax(x + c)` (see https://ogunlao.github.io/2020/04/26/you_dont_really_know_softmax.html).
```

File: src/transformers/tf_utils.py [hunk 3]
```diff
@@ -44,3 +44,22 @@ def shape_list(tensor: Union[tf.Tensor, np.ndarray]) -> List[int]:
...
+
+    Args:
+        logits (`tf.Tensor`). Must be one of the following types: half, float32, float64.
+        axis (`int`, *optional*). The dimension softmax would be performed on. The default is -1 which indicates the
+            last dimension.
+        name (`str`, *optional*). A name for the operation (optional).
```
Comment (sgugger) [inline src/transformers/tf_utils.py:60]:
```suggestion
        name (`str`, *optional*). A name for the operation.
```

**Review by sgugger — APPROVED** [review]
Great find for the bug and thanks a lot for fixing all models!

File: src/transformers/tf_utils.py [hunk 4]
```diff
@@ -44,3 +44,22 @@ def shape_list(tensor: Union[tf.Tensor, np.ndarray]) -> List[int]:
...
+
+    Returns:
+        `tf.Tensor`: A Tensor. Has the same type and shape as logits.
```
Comment (ydshieh) [inline src/transformers/tf_utils.py:63]: nit: guess better to use 

```
`logits`
```
?

Comment (gante) [reply src/transformers/tf_utils.py]: added 👍

File: tests/test_modeling_tf_common.py [hunk 5]
```diff
@@ -1709,6 +1710,40 @@ def call(
...
+
+        # The stable softmax has the same output as the original softmax
+        unstable_out = tf.nn.softmax(masked_x)
+        assert tf.experimental.numpy.allclose(unstable_out, out)
+
+        # We can randomly mask a random numerical input INSIDE XLA
+        xla_out = xla_masked_softmax(x, boolean_mask)
+        out = masked_softmax(x, boolean_mask)
+        assert tf.experimental.numpy.allclose(xla_out, out)
+
```
Comment (ydshieh) [inline tests/test_modeling_tf_common.py:1746]: Could we have a test for bacth_size > 1 ?

**Review by ydshieh — APPROVED** [review]
Good for me, just left 2 nits.
(didn't check the changes in TFGPT2, TFT5 tests though. Let me know if you prefer me to check those too.)

Thank you, @gante  💯

File: tests/test_modeling_tf_common.py (hunk 5 above)
Comment (gante) [reply tests/test_modeling_tf_common.py:1747]: Added batch size > 1 👍

**Comment by Rocketknight1** [comment]
This looks good to me! Do you think it would be better to change `stable_softmax` to only add the offset if we're running on CPU? It makes very little difference either way, but we could hide the complexity of that inside `stable_softmax` and keep our code paths entirely unchanged on GPU. I'm not certain, though - since it's such a small change maybe we can just do it everywhere.

**Review by Rocketknight1 — APPROVED** [review]

**Comment by ydshieh** [comment]
> This looks good to me! Do you think it would be better to change `stable_softmax` to only add the offset if we're running on CPU? It makes very little difference either way, but we could hide the complexity of that inside `stable_softmax` and keep our code paths entirely unchanged on GPU. I'm not certain, though - since it's such a small change maybe we can just do it everywhere.

Good point! Hope this won't affect tests on GPU (at least not for PT/TF equivalence which use `1e-5`). Let's see!

**Comment by gante** [comment]
@Rocketknight1 @ydshieh if you run the test and print the difference between `stable_softmax` and `tf.nn.softmax`, the difference is exactly `0.0` -- I don't think we need to worry about that :D

**Comment by ydshieh** [comment]
@gante With this, do we still have issues regarding sampling in `generate()`. Sorry, I didn't really follow that issue about sampling, but would like to know a bit more 😄

**Comment by gante** [comment]
@ydshieh after this fix, the errors related to `generate()` are gone -- they were caused by the forward pass in the models, which in turn were caused by the issue this PR solves

**Comment by ydshieh** [comment]
(I might be completely wrong below)

I could imagine that we (will) have tests like:

- testing non-XLA and XLA `generte()` that use sampling
  - even with this PR, the differences of output logits between these two might still be as large as, say, `1e-3`?
  - if so, the sampling might give different sampling results ..?
  - if not, what's the magnitude of the diff we get after this PR?
 - testing PT and TF `generte()` that use sampling
   - so same potential issue as above ..? 

Thanks 🙏

**Comment by ydshieh** [comment]
OK, I saw your previous comment

```
I've spun up an Nvidia T4 ( = no tf32 format) and got an error < 1e-5 for all cases
```

**Comment by Rocketknight1** [comment]
Based on the testing results, I'm happy for this to be merged now! If this is an XLA bug, though, we should make sure to revert our changes once none of the TF versions we support are affected by it anymore.

Should we add a TODO to the `masked_softmax` function or a reminder somewhere to make sure that we document why this change is here, and when it can be removed?

**Review by patrickvonplaten — APPROVED** [review]
Great - the solution is clean!

**Comment by gante** [comment]
@Rocketknight1 added a TODO with instructions related to when to deprecate 👍
